webapp/app/sw.index.js

```diff
@@ -59,8 +59,36 @@ self.addEventListener('sync', (event) => {
   }
 });
 
+self.addEventListener('fetch', (event) => {
+  const { request } = event;
+
+  const response = caches
+    .match(request)
+    .then((response) => {
+      if (response) {
+        return response;
+      }
+
+      return fetch(request)
+        .then(resp => {
+          if (request.url.endsWith('.png')) {
+            return caches.open('v1').then(cache => {
+              cache.put(event.request, resp.clone());
+              return resp;
+            });
+          }
```

Comment:
tutaj brakuje obsługi requestów nie kończących się na `.png` i dlatego się wywalało, bo nie potrafił fetchować `index.html`. Trzeba dodać:

`
} else {
    return resp;
}
`

Reply:
OK. Ale w zasadzie dlaczego SW miałby fetchować index.html? Przecież nie cachowaliśmy innych plików statycznych oprócz .png. Trochę nie rozumiem.

Reply:
To działa w taki sposób:

- jak w ramach SW dodajesz event listener na 'fetch' to callback jest odpalany przy każdym requeście
- `caches.match` działa w taki sposób, że do `then` trafia w parametrze pierwszy hit z cache'a jeśli coś się dopausje, lub `undefined` jeśli nic, więc co za tym idzie, dla każdego requesta, zawsze trafiamy do tego `then`'a i jeśli fetchujemy coś co nie jest nigdzie wcześniej cache'owane to pomijamy pierwszego ifa i robimy oryginalnego fetch'a
- returny w promise'ach działają bąbelkowo, to co zwrócisz z promise'a zagnieżdżonego jest zwracane przez promise zewnętrzny
- jeśli po fetchu request nie kończy się na `.png` to nic nie zwracaliśmy z tego promise'a, a więc  cały `return fetch.then(...)` zwracało undefined, a więc `const response = caches` było undefined, a więc robiliśmy `event.respondWith(undefined);` co wywalało całą apkę